### PR TITLE
DEPR: datetimelike.astype(int)

### LINKED ISSUE #24381: Datetimelike index / arrays ignore integer sign and size in astype
For datetimelike indexes, `.astype` ignore the sign and size of an integer dtype, and always uses int64

```python
In [1]: import pandas as pd

In [2]: idx = pd.date_range('2000', periods=4)

In [3]: idx.astype('uint16')
Out[3]:
Int64Index([946684800000000000, 946771200000000000, 946857600000000000,
            946944000000000000],
           dtype='int64')
```

We don't have Index subclasses for each of these types (though we do have uint64 now).

In #24024 we agreed(?) that following the Index was the best behavior for the array class.

```python
In [7]: idx._data.astype('uint16')
Out[7]:
array([946684800000000000, 946771200000000000, 946857600000000000,
       946944000000000000])
```

(that's also int64).

---

Some questions:

1. Do we want to change this? Note that for "normal" usage of datetimes, this is *going* to truncate / overflow. This is particularly unfriendly to 32-bit systems, where "int" is "int32".

```python
In [9]: idx._data.astype('int32').astype("int32")
Out[9]: array([1380122624, -476971008, 1960902656,  103809024], dtype=int32)
```

So maybe we would want safe casting?

```pytb
In [10]: idx._data.astype('int32').astype("int32", casting="safe")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-2c2a4a677a5c> in <module>
----> 1 idx._data.astype('int32').astype("int32", casting="safe")

TypeError: Cannot cast array from dtype('int64') to dtype('int32') according to the rule 'safe'
```

2. ExtensionIndex classes are in the not too distant future, so we may be able to relatively easily get Index classes that work with these different integer sizes / signs. So we might be able to properly support this in the future (if we want to).

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #24381
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Sits on top of #38535

### CODE REVIEW

**Comment by jreback** [comment]
can you merge master

File: doc/source/whatsnew/v1.3.0.rst [hunk 1]
```diff
@@ -142,7 +142,7 @@ Other API changes
 
 Deprecations
 ~~~~~~~~~~~~
-
+- Deprecated casting of datetimelike (``timedelta64[ns]``, ``datetime64[ns]``, ``Datetime64TZDtype``, ``PeriodDtype``) to integer dtypes, use ``values.view(...)`` instead (:issue:`38544`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:145]: casting -> astype

File: pandas/tests/indexes/period/test_astype.py [hunk 2]
```diff
@@ -48,15 +49,17 @@ def test_astype_conversion(self):
         tm.assert_index_equal(result, expected)
 
         idx = period_range("1990", "2009", freq="A", name="idx")
-        result = idx.astype("i8")
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```
Comment (jreback) [inline pandas/tests/indexes/period/test_astype.py:52]: for a lot of these you can just change the test if you want to use view

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.0.rst (hunk 1 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.0.rst:145]: will update

Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.0.rst]: updated+green

File: pandas/tests/indexes/period/test_astype.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/period/test_astype.py]: changed one of these, the rest were specifically targeting astype and not worth mucking with

**Comment by jreback** [comment]
do we have sufficient `.view` tests on *Index* ?

**Comment by jbrockmendel** [comment]
> do we have sufficient .view tests on Index ?

looks decent, yah

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
I ran into this case of datetime->integer casting while exploring our casting rules. @jbrockmendel do you remember the rationale for deprecating this casting? It's not really clear to me why we would in the future disallow datetimelike->integer casting.

This PR doesn't seem to give much reasoning, and the referenced issue is about the bitwidth / sign being ignored, but not sure deprecating altogether is necessarily the fix for that (the bitwidth being ignored *also* applies for integer->integer casting, which we didn't deprecate).

**Comment by jbrockmendel** [comment]
https://github.com/pandas-dev/pandas/issues/22384#issuecomment-909660696

>> We currently raise a warning about the dt64 -> int64 that this will raise instead in the future and to use view instead. Do you know what the rationale was for this deprecation?

> The idea is that 'astype' in general means something like 'the same information but in a different format' and that doesn't make sense with i8 <-> M8

**Comment by jorisvandenbossche** [comment]
Whoops, sorry, I totally forgot I already asked / you answered in the other thread .. :) Will continue there

**Pull request merged**